### PR TITLE
Prepend spaces to nextline elements

### PR DESCRIPTION
**Pull request by josh-huridocs**
#20 

- usage of range.cloneContents
- adding of span element when next element is on the next line
- ctrl+c keyboard event will use the processed text for consistency

### CODE REVIEW

**Comment by Zasa-san** [comment]
I've found an odd problem. When using "click to fill" on a document with the _current_ text selection algorithm I get a pretty good selection from the text as can be seen on the screenshot:

image

But when selecting text with the changes of this branch with click to fill I get the following result:

image

Text seems to be duplicating. I get that same duplication when doing keyboard copy and then pasting on a text document. Right click copy and paste does not present this problem.

File: src/HandleTextSelection.tsx [hunk 1]
```diff
@@ -51,10 +52,29 @@ const HandleTextSelection: FunctionComponent<SelectionHandlerProps> = ({
...
+    const text = getSelectedText();
+    onSelect({ text, selectionRectangles });
   };
 
+  useEffect(() => {
+    const handleKeyDown = (event: KeyboardEvent) => {
+      if ((event.ctrlKey || event.metaKey) && event.key === 'c') {
+        event.preventDefault();
+        const selection = window.getSelection();
+        if (selection && selection.toString().trim() && ref.current) {
```
Comment (Zasa-san) [inline src/HandleTextSelection.tsx:64]: Why do you need to check that the ref exists here?

File: src/HandleTextSelection.tsx [hunk 2]
```diff
@@ -51,10 +52,29 @@ const HandleTextSelection: FunctionComponent<SelectionHandlerProps> = ({
...
+          const text = getSelectedText();
+          navigator.clipboard.writeText(text);
+        }
+      }
+    };
+
+    document.addEventListener('keydown', handleKeyDown);
```
Comment (Zasa-san) [inline src/HandleTextSelection.tsx:71]: Would this not affect every event on the document? Even those that happen outside the text selection area (the pdf document)? Would this not affect a copy event that happens on the sidepanel for example?

File: src/getSelectedText.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  const elementsSelected = prepareSelectedElements();
+  if (elementsSelected.length <= 1) {
+    const selection = window.getSelection();
+    return (selection && selection.toString()) || '';
+  }
+
+  const selectedText = elementsSelected.reduce((acc, el) => {
+    let newText = acc;
+    newText += el.innerText;
+    if (el.tagName === 'BR') {
```
Comment (Zasa-san) [inline src/getSelectedText.ts:43]: I don't follow the point of replacing html line breaks with carry returns. Does that not have the same effect in the end?

File: src/getSelectedText.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  return parseFloat(left1) > parseFloat(left2);
+};
+
+const prepareSelectedElements = (): HTMLElement[] => {
+  const selection = window.getSelection();
+  if (!selection || !selection.rangeCount) return [];
+
+  const range = selection.getRangeAt(0);
+  const selectedContent = range.cloneContents();
+  const selectedElements = selectedContent.querySelectorAll('*');
```
Comment (Zasa-san) [inline src/getSelectedText.ts:14]: You can cast the type here as `NodeListOf<HTMLElement>`

**Review by mfacar — APPROVED** [review]
LGTM! Nice job!

File: src/getSelectedText.ts (hunk 3 above)
Comment (josh-huridocs) [reply src/getSelectedText.ts:43]: br elements when window.getSelection().toString() (the old implementation) is invoked is carry returns

**Review by Zasa-san — CHANGES_REQUESTED** [review]
Nice work! I would suggest that we try to add some test coverage to the changes? Unless you don't think it's necessary.

**Pull request merged**